File: app/src/main/java/com/vo1d/schedulemanager/v2/ui/subjects/list/SubjectKeyProvider.java
package com.vo1d.schedulemanager.v2.ui.subjects.list;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.selection.ItemKeyProvider;
import androidx.recyclerview.widget.RecyclerView;

class SubjectKeyProvider extends ItemKeyProvider<Long> {

    private SubjectsListAdapter adapter;
    private RecyclerView recyclerView;

    public SubjectKeyProvider(SubjectsListAdapter adapter, RecyclerView recyclerView) {
        super(SCOPE_MAPPED);
        this.adapter = adapter;
        this.recyclerView = recyclerView;
    }


    @Nullable
    @Override
    public Long getKey(int position) {
        return adapter.getItemId(position);
    }

    @Override
    public int getPosition(@NonNull Long key) {
        RecyclerView.ViewHolder viewHolder = recyclerView.findViewHolderForItemId(key);
        return viewHolder == null ? RecyclerView.NO_POSITION : viewHolder.getAdapterPosition();
    }
}

File: app/src/main/kotlin/ru/vo1dmain/ttmanager/sections/subjects/list/SubjectKeyProvider.kt
package ru.vo1dmain.ttmanager.sections.subjects.list

import androidx.recyclerview.selection.ItemKeyProvider
import androidx.recyclerview.widget.RecyclerView

internal class SubjectKeyProvider(
    private val adapter: SubjectsListAdapter,
    private val recyclerView: RecyclerView
) : ItemKeyProvider<Long>(SCOPE_MAPPED) {
    
    override fun getKey(position: Int) =
        adapter.getItemId(position)
    
    override fun getPosition(key: Long): Int {
        val viewHolder = recyclerView.findViewHolderForItemId(key)
        return viewHolder?.adapterPosition ?: RecyclerView.NO_POSITION
    }
}